ledger/ledger-on-posix-filesystem/src/main/scala/com/daml/ledger/on/filesystem/posix/FileSystemLedgerReaderWriter.scala

```diff
@@ -0,0 +1,216 @@
+    lock.run {
+      for {
+        stateInputStream <- Future.sequence(
+          submission.getInputDamlStateList.asScala.toVector
+            .map(key => readState(key).map(key -> _)))
+        stateInputs: Map[DamlStateKey, Option[DamlStateValue]] = stateInputStream.toMap
+        entryId = allocateEntryId()
+        (logEntry, stateUpdates) = KeyValueCommitting.processSubmission(
+          engine,
+          entryId,
+          currentRecordTime(),
+          LedgerReader.DefaultTimeModel,
+          submission,
+          participantId,
+          stateInputs
+        )
+        _ = verifyStateUpdatesAgainstPreDeclaredOutputs(stateUpdates, entryId, submission)
+        newHead <- appendLog(entryId, Envelope.enclose(logEntry))
+        _ <- updateState(stateUpdates)
+      } yield {
+        dispatcher.signalNewHead(newHead)
+        SubmissionResult.Acknowledged
+      }
+    }
+  }
+
+  private def verifyStateUpdatesAgainstPreDeclaredOutputs(
+      actualStateUpdates: Map[DamlStateKey, DamlStateValue],
+      entryId: DamlLogEntryId,
+      submission: DamlSubmission
+  ): Unit = {
+    val expectedStateUpdates = KeyValueCommitting.submissionOutputs(entryId, submission)
+    if (!(actualStateUpdates.keySet subsetOf expectedStateUpdates)) {
+      val unaccountedKeys = actualStateUpdates.keySet diff expectedStateUpdates
+      sys.error(
+        s"CommitActor: State updates not a subset of expected updates! Keys [$unaccountedKeys] are unaccounted for!")
+    }
+  }
+
+  private def currentRecordTime(): Timestamp =
+    Timestamp.assertFromInstant(Clock.systemUTC().instant())
+
+  private def allocateEntryId(): DamlLogEntryId = {
+    val nonce: Array[Byte] = Array.ofDim(8)
+    randomNumberGenerator.nextBytes(nonce)
+    DamlLogEntryId.newBuilder
+      .setEntryId(ByteString.copyFrom(nonce))
+      .build
+  }
+
+  private def retrieveLogEntry(index: Index): Future[LedgerRecord] =
+    for {
+      entryId <- Future(Files.readAllBytes(logIndexDirectory.resolve(index.toString))).transform {
+        case Success(value) => Success(value)
+        case Failure(exception) if exception.isInstanceOf[NoSuchFileException] =>
+          Failure(new NoSuchElementException(s"No log entry at $index."))
+        case Failure(exception) => Failure(exception)
+      }
+      envelope <- Future(Files.readAllBytes(logEntriesDirectory.resolve(Bytes.toString(entryId))))
+    } yield
+      LedgerRecord(
+        Offset(Array(index.toLong)),
+        DamlLogEntryId.newBuilder().setEntryId(ByteString.copyFrom(entryId)).build(),
+        envelope)
+
+  private def appendLog(entry: DamlLogEntryId, envelope: ByteString): Future[Index] =
+    for {
+      currentHead <- Future(Files.readAllLines(logHeadPath).get(0)).transform {
+        case Success(contents) =>
+          Success(contents.toInt)
+        case Failure(exception) if exception.isInstanceOf[NoSuchFileException] =>
```

Comment:
This could be simplified as
```suggestion
        case Failure(_: NoSuchFileException) =>
```

Reply:
Done. Thanks for the tip!